File: src/main/python/modular/compute/MDSolver.py
from typing import Any, Optional
from collections import defaultdict

import networkx as nx

from modular.tree.RootedForest import RootedForest, Node
from modular.MDNode import MDNode, OperationType
from modular.compute.MDComputeNode import MDComputeNode, NodeType
from modular.compute.pivot import process_neighbors, do_pivot
from modular.compute.misc import remove_extra_components, remove_layers, complete_alpha_lists, merge_components
from modular.compute.refinement import refine
from modular.compute.promotion import promote
from modular.compute.assembly import assemble

VertexId = Any


def trace(message: str, *args: object) -> None:
    print(f'[TRACE] {message}', *args)


def compute(
    G: nx.Graph,
    tree: RootedForest[MDComputeNode],
    vertex_nodes: dict[VertexId, Node[MDComputeNode]],
    main_prob: Node[MDComputeNode]
) -> Node[MDComputeNode]:
    current_prob: Optional[Node[MDComputeNode]] = main_prob

    alpha_list: dict[VertexId, set[VertexId]] = defaultdict(set)
    visited: set[VertexId] = set()
    result = None
    t = 0

    while current_prob:
        t += 1
        # trace(f'solve start: current ({t}): {current_prob}, tree={tree}, visited={visited}, alpha={dict(alpha_list)}')
        fc = current_prob.first_child
        assert fc is not None

        current_prob.data.active = True

        if fc.data.node_type != NodeType.PROBLEM:
            # first, needs to solve subproblems
            visited.add(fc.data.vertex)

            if current_prob.has_only_one_child():
                # base case
                process_neighbors(G, tree, vertex_nodes, alpha_list, visited, fc.data.vertex, current_prob, None)
            else:
                # pivot at the first child
                pivoted = do_pivot(G, tree, vertex_nodes, alpha_list, visited, current_prob, fc.data.vertex)

                # dig into the first subproblem
                assert pivoted.first_child is not None
                current_prob = pivoted.first_child
                continue
        else:
            # print(f'[TRACE] solve middle: current ({t}): {current_prob}, tree={tree}, visited={visited}, alpha={dict(alpha_list)}')
            # now, ready to compute this problem
            extra_components = remove_extra_components(tree, current_prob)

            remove_layers(tree, current_prob)
            complete_alpha_lists(current_prob, alpha_list)
            # trace(f'alpha={dict(alpha_list)}, current={tree}')
            refine(tree, vertex_nodes, alpha_list, current_prob)
            # trace(f'after refine: current={current_prob}')
            promote(tree, current_prob)
            # trace(f'after promote: current={current_prob}')
            assemble(tree, vertex_nodes, alpha_list, current_prob)
            # trace(f'after assemble: current={current_prob}')
            merge_components(tree, current_prob, extra_components)
            # trace(f'after merge_components: current={current_prob}')

            # clear all but visited
            assert current_prob.first_child is not None
            for c in current_prob.first_child.dfs_reverse_preorder_nodes():
                if c.is_leaf() and c.data.vertex in alpha_list:
                    del alpha_list[c.data.vertex]
                c.data.clear()

        # trace(f'solve finish: current ({t}): {current_prob}, tree={tree}, visited={visited}, alpha={dict(alpha_list)}')
        result = current_prob.first_child
        current_prob = current_prob.parent if current_prob.is_last_child() else current_prob.right

        assert len(tree.roots) == 1, f'unclean tree: {tree.roots}'

    # set new root of the tree
    assert result is not None
    result_parent = result.parent
    assert result_parent is not None

    tree.detach(result)
    tree.remove(result_parent)
    return result


class MDSolver:
    @staticmethod
    def compute(G: nx.Graph) -> tuple[RootedForest[MDNode], Node[MDNode], list[VertexId]]:
        n = len(G)
        assert n > 0, 'empty graph'

        # Convert node names to consecutive integers starting from 0.
        # The original labels are preserved as the `label` attribute for each node.
        G = nx.convert_node_labels_to_integers(G, first_label=0, ordering='sorted', label_attribute='label')
        labels = nx.get_node_attributes(G, 'label')

        # build computation tree
        tree: RootedForest[MDComputeNode] = RootedForest()

        # create the main problem
        main_prob = tree.create_node(MDComputeNode.new_problem_node(connected=False))

        # add vertex nodes (cannot be removed)
        vertex_nodes = {}
        for v in reversed(list(G.nodes())):
            node = tree.create_node(MDComputeNode.new_vertex_node(v))
            vertex_nodes[v] = node
            tree.move_to(node, main_prob)

        # main logic
        comp_root = compute(G, tree, vertex_nodes, main_prob)

        # create result tree
        result_tree: RootedForest[MDNode] = RootedForest()
        mapping: dict[Node[MDComputeNode], Node[MDNode]] = {}
        vertices = []

        # create vertex nodes in DFS ordering (right-to-left)
        for i, comp_node in enumerate(comp_root.get_leaves()):
            v = labels[comp_node.data.vertex]
            vertices += [v]
            nd = result_tree.create_node(MDNode(v, OperationType.PRIME, i, i + 1))
            mapping[comp_node] = nd

        # create internal nodes from the bottom
        for comp_node in reversed(list(comp_root.bfs_nodes())):
            if comp_node.data.is_vertex_node():
                continue
            assert comp_node.data.is_operation_node(), 'there should not be a problem node'

            children = comp_node.get_children()
            idx_begin = min(mapping[c].data.vertices_begin for c in children)
            idx_end = max(mapping[c].data.vertices_end for c in children)
            nd = result_tree.create_node(MDNode(op=comp_node.data.op_type, vertices_begin=idx_begin, vertices_end=idx_end))
            for c in reversed(children):
                result_tree.move_to(mapping[c], nd)
            mapping[comp_node] = nd

        return result_tree, mapping[comp_root], vertices

File: src/main/cpp/modular/compute/MDSolver.cpp
#include "MDSolver.hpp"

namespace modular {
namespace compute {
namespace impl {

int compute(ds::graph::Graph const &graph, CompTree &tree, int main_prob, util::Profiler *prof) {
  TRACE("start compute(): %s", tree.to_string(main_prob).c_str());
  PROF(util::pstart(prof, "compute()"));

  int n = graph.number_of_nodes();
  int current_prob = main_prob;

  VVV alpha_list(n);
  VVV fp_neighbors(n);  // used only for assembly -> compute_fact_perm_edges()
  bool visited[n];
  for (int i = 0; i < n; ++i) {
    visited[i] = false;
  }
  ds::FastSet vset(n);
  int result = -1;
  int t = 0;

  while (tree.is_valid(current_prob)) {
    ++t;
    TRACE("main problem (%d): %s", t, tree.to_string(tree.get_root(current_prob)).c_str());
    TRACE("current problem: %s", tree.to_string(current_prob).c_str());
    for (int i = 0; i < n; ++i) {
      TRACE("visited [%d]: %s", i, visited[i] ? "True" : "False");
      TRACE("alpha [%d]: %s", i, util::to_string(alpha_list[i]).c_str());
    }

    auto &cp = tree[current_prob];
    cp.data.active = true;

    auto &fc = tree[cp.first_child];

    if (!fc.data.is_problem_node()) {
      // first, needs to solve subproblems
      visited[cp.first_child] = true;  // set visited

      if (cp.has_only_one_child()) {
        // base case
        PROF(util::pcount(prof, "solve(): base case"));
        PROF(util::pstart(prof, "process_neighbors()"));
        process_neighbors(graph, tree, alpha_list, visited, cp.first_child, current_prob, -1);
        PROF(util::pstop(prof, "process_neighbors()"));
      } else {
        // pivot at the first child
        PROF(util::pstart(prof, "do_pivot()"));
        auto pivoted = do_pivot(graph, tree, alpha_list, visited, current_prob, cp.first_child);
        PROF(util::pstop(prof, "do_pivot()"));

        // dig into the first subproblem
        current_prob = tree[pivoted].first_child;
        continue;
      }
    } else {
      // now, ready to compute this problem
      PROF(util::pstart(prof, "remove_extra_components()"));
      auto extra_components = remove_extra_components(tree, current_prob);
      PROF(util::pstop(prof, "remove_extra_components()"));
      TRACE("extra: %s", tree.to_string(extra_components).c_str());

      PROF(util::pstart(prof, "remove_layers()"));
      remove_layers(tree, current_prob);
      PROF(util::pstop(prof, "remove_layers()"));

      PROF(util::pstart(prof, "complete_alpha_lists()"));
      std::vector<int> leaves = tree.get_leaves(current_prob);
      complete_alpha_lists(tree, alpha_list, vset, current_prob, leaves);
      PROF(util::pstop(prof, "complete_alpha_lists()"));

      PROF(util::pstart(prof, "refine()"));
      refine(tree, alpha_list, current_prob, leaves, prof);
      PROF(util::pstop(prof, "refine()"));

      PROF(util::pstart(prof, "promote()"));
      promote(tree, current_prob);
      PROF(util::pstop(prof, "promote()"));

      PROF(util::pstart(prof, "assemble()"));
      assemble(tree, alpha_list, current_prob, fp_neighbors, vset, prof);
      PROF(util::pstop(prof, "assemble()"));

      // clear all but visited
      PROF(util::pstart(prof, "clear all but visited"));
      for (auto c: tree.dfs_reverse_preorder_nodes(tree[current_prob].first_child)) {
        if (tree[c].is_leaf()) alpha_list[c].clear();
        tree[c].data.clear();
      }
      PROF(util::pstop(prof, "clear all but visited"));

      PROF(util::pstart(prof, "merge_components()"));
      merge_components(tree, current_prob, extra_components);
      PROF(util::pstop(prof, "merge_components()"));
    }

    result = tree[current_prob].first_child;
    current_prob = tree[current_prob].is_last_child() ? tree[current_prob].parent : tree[current_prob].right;
    TRACE("loop: %s", tree.to_string(result).c_str());
  }

  // set new root of the tree
  auto result_parent = tree[result].parent;
  tree.detach(result);
  tree.remove(result_parent);

  PROF(util::pstop(prof, "compute()"));
  TRACE("return: %s", tree.to_string(result).c_str());
  return result;
}
}  // namespace impl
}  // namespace compute
}  // namespace modular
